### ISSUE TITLE
Maybe.of vs toMaybe

### ISSUE THREAD

**Issue by marcusnielsen** [TylorS/typed#18]
I was looking through the source code and saw that `Maybe.of` is an alias for `Just.of`, and `toMaybe` does a null check before either returning a `Nothing`, or passing the result to `Maybe.of`/`Just.of`.

I was expecting `Maybe.of` to also do the null check and return a `Nothing` or a `Just`.

Is there a standard way of thinking about `of` and `toXYZ` in these kinds of datatypes in FP in general?

**Comment by TylorS** [comment]
Hey @marcusnielsen,

So you're not the first person to find confusion in having both `toMaybe` and `Maybe.of`, and the second time at least where someone has expected `Maybe.of` to have the behavior that only `toMaybe` has. `Maybe.of` was originally made only as a convenience because you'll often have to import `Maybe` for the _type_ already.

> Is there a standard way of thinking about of and toXYZ in these kinds of datatypes in FP in general
I try to think about it pretty literally, but I don't think there is a hard rule on these things. Across the different `@typed/*` packages I've used the `DataStructure.of` API as a way to create some _of_ that type. To create a `Maybe ` _of_ any type of value it's a `Just` containing that value since `Just<A>` is a valid `Maybe<A>`.  

`toMaybe` was originally meant to deal with 3rd party APIs that originally work with `A | void` style functions to signal failures.

```typescript
// external-lib.ts
function foo<A>(value: A): A | void { ... }

// bar.ts
import { foo } from 'external-lib'
import { pipe } from '@typed/functions'
import { toMaybe } from '@typed/maybe'

export const bar: <A>(value: A): Maybe<A> = pipe(foo, toMaybe)
```

**Comment by Frikki** [comment]
I think the `DataStructure.of` is semantically correct. `toMaybe` is perhaps less clear. Perhaps (I’m purposely avoiding the word _maybe_ here) `asMaybe`, `inMaybe` or `wrapWithMaybe` would be clearer?

Is there any confusion about `Just` and `Nothing` being `Maybe`s?

**Comment by marcusnielsen** [comment]
My confusion is rooted in `Maybe.of`, and `Just.of`.

`Just.of` can only give a `Just` back. Very clear.
`Maybe` could be a `Just` or a `Nothing`, so it would make sense for `of` to convert it to either a `Just` or `Nothing`, since there's already a `Just.of` for wrapping in a `Just`.

I just think that the API should be consistent. It sounds like the `of` methods just wrap a structure around some value in the simplest way possible, while `toDataStructure` functions do some smarter convertion.

It makes sense! Even though there's a lot of beginners not getting it straight away, it might still be correct to keep it as it is. At least it feels like it makes sense to me now.

Closing the issue since I feel it got resolved, but you can add to the discussion if you want.

Thanks for your help @TylorS and @Frikki!

**Comment by TylorS** [comment]
This confusion has been addressed in v6.0.0 of `@typed/maybe`. `Maybe.of` now has the behavior previously only found in `toMaybe` and `toMaybe` itself has been removed from the API.

https://github.com/TylorS/typed/tree/master/packages/maybe#maybeofavalue-a-maybea

**Comment by marcusnielsen** [comment]
That sounds great! :trophy:
